feat(ArrayField): enhance add method to support optional index-based insertion issue#2544

Shabi-x opened the pull request:
<!-- Thanks so much for your PR 💗 -->
[中文模板 / Chinese Template](https://github.com/DouyinFE/semi-design/blob/main/.github/PULL_REQUEST_TEMPLATE.zh-CN.md)

- [ ] I have read and followed [Pull Request Guidelines](https://github.com/DouyinFE/semi-design/blob/main/CONTRIBUTING-en-US.md#pull-request-guidelines) of the contributing guide.


### What kind of change does this PR introduce? (check at least one)

 - [ ] Bugfix
 - [x] Feature
 - [ ] Code style update
 - [ ] Refactor
 - [x] Test Case
 - [ ] TypeScript definition update
 - [ ] Document improve
 - [ ] CI/CD improve
 - [ ] Branch sync
 - [ ] Other, please describe:


### PR description
<!--
The relevant issue, background of this PR, and what should reviewers focus on
-->
Fixes #

### Changelog
🇨🇳 Chinese
-feat: 根据 issue#2544的建议，针对arrayfiled组件add只能始终添加到最后一个位置，我阅读代码后修改了一下add方法，提供了一个可选的index来给与使用者添加新项位置的可选性，并添加了一个storybook测试项进行了一些测试。

---

🇺🇸 English
- feat(ArrayField): enhance add method to support optional index-based insertion related to issue #2544 
- 


### Checklist
- [x] Test or no need
- [ ] Document or no need
- [ ] Changelog or no need

### Other
- [ ] Skip Changelog

### Additional information
<!-- You can provide screenshot/video or some additional information -->

Commit:
feat(form/ArrayField): enhance add method to support optional index-based insertion

Shabi-x commented:
在做一些开源的尝试，有不足之处欢迎批评指正，谢谢！

Shabi-x commented:
@pointhalo 麻烦您了 请问这样解决这个issue可以嘛 希望得到您的批评指点~

pointhalo commented:
抱歉，最近事情比较多。处理速度没跟上，预计下周整体看一下
- 另外可以先补充一下对应逻辑的测试用例，jest unit test 或者 cypress e2e test都可以

Shabi-x commented:
> 抱歉，最近事情比较多。处理速度没跟上，预计下周整体看一下
> 
> * 另外可以先补充一下对应逻辑的测试用例，jest unit test 或者 cypress e2e test都可以

好的，我看一下哈

Commit:
test(arrayfield): add comprehensive tests for indexed array field insertion

Shabi-x reviewed (commented):
新增的测试用例通过，但arrayfield.test.js的一项遇到了问题，我会排查一下错误根源

Shabi-x commented:
> 新增的测试用例通过，但arrayfield.test.js的一项遇到了问题，我会排查一下错误根源

之前的做法是希望用户使用add时默认在最后一行插入新行，用户可以     
                                     <Button onClick={add} type="primary">
                                        Add
                                    </Button>这样调用add方法，
调整add方法后，支持传入索引了    

add(index?: number) {
        const { keys } = this.state;
        const { field } = this.props;
        const updater = this.context;
        const newKey = getUuidv4();

        if (index === undefined) {
            keys.push(newKey);
        } else {
            keys.splice(index + 1, 0, newKey);
        }

        this.shouldUseInitValue = true;
        this.setState({ keys });

        let updateKey = new Date().valueOf();
        updater.updateArrayField(field, { updateKey });
        this.cacheUpdateKey = updateKey;

        return newKey;
    }

                                    但是这样修改之后，用户用        
                                    <Button onClick={() => add()} type="primary">
                                        Add
                                    </Button>箭头函数的方式调用，表现正常。
                                    但是不兼容直接             
                                    <Button onClick={add} type="primary">
                                        Add
                                    </Button>的调用方法了

所以测试没有通过

Shabi-x commented:
> > 新增的测试用例通过，但arrayfield.test.js的一项遇到了问题，我会排查一下错误根源
> 
> 之前的做法是希望用户使用add时默认在最后一行插入新行，用户可以传入索引在任意位置插入，未传入索引，默认插在最后一行。由于我的测试代码都是箭头函数调用，没有出现问题。之前的demo出现了如下的问题：
> 
> 调整add方法后，支持传入索引
> 
> ```
> add(index?: number) {
>         const { keys } = this.state;
>         const { field } = this.props;
>         const updater = this.context;
>         const newKey = getUuidv4();
> 
>         if (index === undefined) {
>             keys.push(newKey);
>         } else {
>             keys.splice(index + 1, 0, newKey);
>         }
> 
>         this.shouldUseInitValue = true;
>         this.setState({ keys });
> 
>         let updateKey = new Date().valueOf();
>         updater.updateArrayField(field, { updateKey });
>         this.cacheUpdateKey = updateKey;
> 
>         return newKey;
>     }
> ```
> 
> 但是这样修改之后，用户用
> 
> ```
>  <Button onClick={() => add()} type="primary">
>       Add
>  </Button>
> ```
> 
> 箭头函数的方式调用，表现正常。 但是不兼容直接
> 
> ```
> <Button onClick={add} type="primary">
>      Add
> </Button>
> ```
> 
> 的调用方法了 所以测试没有通过，这里还需要寻找问题并修改

Shabi-x commented:
> > > 新增的测试用例通过，但arrayfield.test.js的一项遇到了问题，我会排查一下错误根源
> > 
> > 
> > 之前的做法是希望用户使用add时默认在最后一行插入新行，用户可以传入索引在任意位置插入，未传入索引，默认插在最后一行。由于我的测试代码都是箭头函数调用，没有出现问题。之前的demo出现了如下的问题：
> > 调整add方法后，支持传入索引
> > ```
> > add(index?: number) {
> >         const { keys } = this.state;
> >         const { field } = this.props;
> >         const updater = this.context;
> >         const newKey = getUuidv4();
> > 
> >         if (index === undefined) {
> >             keys.push(newKey);
> >         } else {
> >             keys.splice(index + 1, 0, newKey);
> >         }
> > 
> >         this.shouldUseInitValue = true;
> >         this.setState({ keys });
> > 
> >         let updateKey = new Date().valueOf();
> >         updater.updateArrayField(field, { updateKey });
> >         this.cacheUpdateKey = updateKey;
> > 
> >         return newKey;
> >     }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 但是这样修改之后，用户用
> > ```
> >  <Button onClick={() => add()} type="primary">
> >       Add
> >  </Button>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 箭头函数的方式调用，表现正常。 但是不兼容直接
> > ```
> > <Button onClick={add} type="primary">
> >      Add
> > </Button>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 的调用方法了 所以测试没有通过，这里还需要寻找问题并修改

经过debug发现当onclick={add}时的index并不是undifined，而是
![image](https://github.com/user-attachments/assets/ccd2f456-1bde-49b1-8a0f-84b73ca0d673)

是处理边界的问题，这边对index的类型加以判断，修正了这个问题：）
![image](https://github.com/user-attachments/assets/95cd474f-4aa4-4362-bf4e-536ebe6df1dc)
@pointhalo

Commit:
feat(ArrayField): improve index type boundary handling for add method

Commit:
feat(ArrayField): improve index type boundary handling for add method

Commit:
Merge branch 'feat-shabix-2544' of github.com:Shabi-x/semi-design into feat-shabix-2544

Commit:
test(ArrayField): Enhance test coverage for array insertion methods

SudoUserReal commented:
感谢提交 PR。整体看了一下改动，有几个问题需要处理：

---

### 1. 插入时未同步更新 Form State 的 values 和 errors（需修复）

这个是主要问题。当前 `add(index)` 只在 `keys` 数组中插入了新 key，没有同步更新 Form State 中的 values 和 errors 数组。

`render` 中字段 path 是按 keys 数组索引生成的（`tasks[0]`, `tasks[1]`, ...），在非末尾位置插入新 key 后，原有 keys 的索引会右移，但 Form State 中的 values 没有对应移位，会导致已有字段数据错位。

例如：
- 插入前：keys = [k1, k2]，form values = `{tasks: ["值1", "值2"]}`
- `add(0)` 后：keys = [k3, k1, k2]
- 渲染时：k3 → `tasks[0]`（拿到"值1"），k1 → `tasks[1]`（拿到"值2"），k2 → `tasks[2]`（空）

可以参考 `remove` 方法的做法，`remove` 在删除 key 的同时也会同步 splice values 和 errors。`add` 在指定 index 插入时需要做对等的处理：

```typescript
if (typeof index === 'number' && index < keys.length) {
    keys.splice(index, 0, newKey);
    let currentValues = updater.getValue(field);
    if (Array.isArray(currentValues)) {
        currentValues = currentValues.slice();
        currentValues.splice(index, 0, undefined);
        updater.updateStateValue(field, currentValues, { notNotify: true, notUpdate: true });
    }
    let currentErrors = updater.getError(field);
    if (Array.isArray(currentErrors)) {
        currentErrors = currentErrors.slice();
        currentErrors.splice(index, 0, undefined);
        updater.updateStateError(field, currentErrors, { notNotify: true, notUpdate: true });
    }
} else {
    keys.push(newKey);
}
```

---

### 2. TypeScript 接口未更新

`ArrayFieldChildrenProps` 中 `add` 的类型还是 `add: () => void`，需要改成 `add: (index?: number) => void`。

---

### 3. 条件判断可以简化

```typescript
if (index === undefined || (index && typeof index !== 'number')) {
```

直接写 `typeof index !== 'number'` 就够了，能覆盖 `undefined`、`onClick={add}` 传入 SyntheticEvent 等所有非数字情况。

---

### 4. `addWithInitValue` 也需要支持 index

为了 API 一致性，`addWithInitValue` 也应该支持 `addWithInitValue(rowVal, index?)` 形式。

---

### 5. `add` 方法缺少 return

Storybook demo 里用了 `const newKey = add(1)`，但方法体里没有 `return newKey`，实际拿到的是 `undefined`。如果要支持返回 key，加一下 return。

---

### 6. 测试覆盖不够

当前测试只验证了 length 变化，没有验证插入位置是否正确、已有字段值在插入后是否保持不变。建议补充带初始值的场景，断言各位置的值。这样也能暴露第 1 点的问题。

---

### 7. 建议增加 index 边界校验

负数 index 传给 splice 会从末尾倒数，可能不是预期行为。建议加一下范围限制，比如 `Math.max(0, Math.min(index, keys.length))`。

---

### 8. API 语义

Issue 原始描述是"在索引下方插入"（insert after），当前实现是 `splice(index, 0, ...)` 即 insert at 语义。这个和 Ant Design Form.List 的 add 一致，没问题，但文档和 demo 里的描述需要对齐，明确 `add(index)` 是"在 index 位置插入"。

---

其中第 1 点是阻塞项，其余可以一起改。

SudoUserReal committed:
fix(form): sync form state on index-based insertion in ArrayField

- Sync values and errors arrays when inserting at a specific index,
  preventing data misalignment with existing fields
- Clamp index to valid range (0 ~ keys.length)
- Simplify type guard to typeof index !== 'number'
- Add return newKey from add()
- Support optional index param in addWithInitValue
- Update ArrayFieldChildrenProps type definitions
- Rewrite tests to verify insertion position and value correctness
- Fix storybook demo labels to match insert-at semantics

SudoUserReal commented:
我直接在这个分支上提交了一个 fix commit (6a0ecb01)，主要改动：

**arrayField.tsx**
- `add(index?)` 在指定 index 插入时，同步 splice form state 的 values 和 errors 数组，避免已有字段数据错位（参考 `remove` 方法的对等处理）
- index 边界校验：`Math.max(0, Math.min(index, keys.length))`，防止负数和越界
- 类型判断简化为 `typeof index !== 'number'`
- 增加 `return newKey`
- `addWithInitValue` 新增可选 `index` 参数
- `ArrayFieldChildrenProps` 类型同步更新：`add: (index?: number) => void`、`addWithInitValue: (lineObject, index?) => void`

**测试**
- 重写 arrayfieldinsert.test.js，覆盖：默认追加、头部/中间/尾部插入、越界 clamp、负数 clamp、空数组、addWithInitValue 带/不带 index、onClick={add} 直传事件对象
- 原有 arrayField.test.js 5 个用例全部通过

**storybook demo**
- 按钮文案改为 insert-at 语义，和实际行为对齐
